Mutational signatures

MaximiliaEggle opened the pull request:
Hi all, 

I made a first suggestion on how to include per sample mutational signature (+probability) output from SigAssignment into LACHESIS. 

As SigAssignment needs to be run on maf files (some vcf callers are already supported, but not all) and in python, I think it would be the easiest way to let the user run SigAssignment and specify a filepath to the output. 

Decomposed_MutationType_Probabilities.txt looks like this:

```
Sample Names	MutationType	SBS1	SBS2	SBS3	SBS4	SBS5	SBS6	SBS7a	SBS7b	SBS7c	SBS7d	SBS8	SBS9	SBS10a	SBS10b	SBS10c	SBS10d	SBS11	SBS12	SBS13	SBS14	SBS15	SBS16	SBS17a	SBS17b	SBS18	SBS19	SBS20	SBS21	SBS22a	SBS22b	SBS23	SBS24	SBS25	SBS26	SBS27	SBS28	SBS29	SBS30	SBS31	SBS32	SBS33	SBS34	SBS35	SBS36	SBS37	SBS38	SBS39	SBS40a	SBS40b	SBS40c	SBS41	SBS42	SBS43	SBS44	SBS45	SBS46	SBS47	SBS48	SBS49	SBS50	SBS51	SBS52	SBS53	SBS54	SBS55	SBS56	SBS57	SBS58	SBS59	SBS60	SBS84	SBS85	SBS86	SBS87	SBS88	SBS89	SBS90	SBS91	SBS92	SBS93	SBS94	SBS95	SBS96	SBS97	SBS98	SBS99
B087-NB_12524	A[C>A]A	0.0010425381415898572	0.0	0.0	0.0	0.15865069172701046	0.0	0.0	0.0	0.0	0.0	0.37561323502030164	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.426821811123005	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.005335527377981928	0.0	0.0	0.0	0.0	0.0	0.0	0.032536196610111145	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0	0.0

```

Therefore the code includes adding Sample Names and MutationType as column for nb (to merge later on). For MutationType we need the Trinucleotide sequence (sequence_context column) that can be read from vcf directly for dkfz or is assigned using BSgenome if this column doesn't exist. 
Using the parameters min.p and sig.select the user can specify the minimum probability for signatures to be considered and the signatures to be considered (i.e. only clock-like). 

At plotnb in addition to the normal plot, a stratified VAF histogram with the mutational signatures can be plotted if sig.show = TRUE.
[VAF_histogram_strat.pdf](https://github.com/user-attachments/files/20161306/VAF_histogram_strat.pdf)
[VAF_histogram_strat_sig.pdf](https://github.com/user-attachments/files/20161307/VAF_histogram_strat_sig.pdf)

I will change the colors to stay the same for the patient and the whole cohort and add it. The devtools::check() and test on my cohorts looked fine so far. What do you think about this?

VerenaK90 commented:
Hi Maxi,

thanks again for developing this! I think there's a small bug in the signature assignment function. The file "Decomposed_MutationType_Probabilites.txt" does not contain columns for Chr and Pos. I think we can just assign the probabilities based on sample and substitution type and don't need to assign position-wise (as the probabilities will always be the same for one substitution type). The commits I did try to implement this, using code you developed beforehand and some adjustments. Could you please have a look at this? Probably not everything I did is implemented in an ideal way from a coding perspective.

MaximiliaEggle commented:
Hi Verena, 

I am sorry, I used the name Decomposed_MutationType_Probabilites.txt, but actually meant the per sample output and didn't realize that there is already a merged version without Chr and Pos, this is a little confusing. 
If you run the SigAssignment analysis with input_type=”vcf” (input can be vcf, maf, .txt) instead of "matrix", you will get a per sample Decomposed_MutationType_Probabilites_ID.txt file with Pos and Chr (see NBE15 https://github.com/VerenaK90/LACHESIS/blob/5240eca2ce8ef9fbf536b31e1c20c1e34e00f9e7/inst/extdata/NBE15/Decomposed_Mutation_Probabilities_NBE15.txt). I merged all of these files and used them as an input for cohort analysis. 

I will have a look at the code and edit the vignette/ Roxygen to explain the SigAssignment analysis today.

MaximiliaEggle commented:
Hi all, 

as Verena suggested and implemented we can use the mutation type context file instead of the ID/ per SNV file as they are the same and the user doesn't need to merge them before inputting it. 

I made minor changes (simplifying sv.1 and sv.2 directly as sv) and passed reference.build as a parameter. In addition, I guess we need to use allow.cartesian = TRUE in the merge command, which is slowing down LACHESIS a lot. It is way faster with IRanges and GenomicRanges (#Option1). I know that we don't want to use too many dependencies, @PoisonAlien maybe we could look at this together on Monday and see if you have another idea. 
For reproducability I implemented automatic seed generation or user-specified seed input, that is saved in the logfile. 

Furthermore, I added this to the vignette. I needed to move the Decomposed_MutationType_Probabilities.txt file and changed this in the Roxygen. And RColorBrewer and Biostrings are now listed as dependencies in the DESCRIPTION. 

Best, 
Maxi

VerenaK90 commented:
Thank you, Maxi! I think your example file still uses  Decomposed_MutationType_Probabilities_NBE15.txt and this causes the problem with `allow.cartesian`. I think if you directly use the file Decomposed_MutationType_Probabilities.txt, you don't have to set `allow.cartesian=TRUE`. I don't think there is additional information in Decomposed_MutationType_Probabilities_NBE15.txt as compared to Decomposed_MutationType_Probabilities.txt if we do the assignment based on substitution type rather than position, though I might be mistaken.

MaximiliaEggle commented:
Oh you are right, I totally forgot to change the NBE15 example file. I removed the allow.cartesian = TRUE. Option 1 would be a lot faster (depends on the sample, but i.e. 10seconds instead of 30seconds), but includes IRanges and GenomicRanges as dependencies. For Option 2 I added a few comments, so the pyrimidization becomes a little clearer. Anand was busy yesterday/today, but he said he would look at it today/tomorrow. Maybe he has another idea to make it faster without dependencies.

PoisonAlien commented:
Hi Maxi,

Impressive use of data.table and BSgenomes. My suggestion


1. No need to assign to [SV again](https://github.com/VerenaK90/LACHESIS/blob/0c5f5103661d6f18d02008e65476d62787a5740f/R/nbImport.R#L147). `:=` happens in place 


2. No need for switch statement again at [132](https://github.com/VerenaK90/LACHESIS/blob/0c5f5103661d6f18d02008e65476d62787a5740f/R/nbImport.R#L130). 


```r
switch(genome, ..)
```

3. **Important**: Dont use .I This, which makes it process row-by-row - significantly slows it down. Removing it here for faster vectorized ops. Just [remove .I](https://github.com/VerenaK90/LACHESIS/blob/0c5f5103661d6f18d02008e65476d62787a5740f/R/nbImport.R#L138) Its faster than Granges

4. Please use namespaces for data.table functions. example for [fread here](https://github.com/VerenaK90/LACHESIS/blob/0c5f5103661d6f18d02008e65476d62787a5740f/R/nbImport.R#L99)


5. If we are importing BSgenome objects, Granges will be reverse-imported. If you would like to use the GRanges functions, please do so. There is no avoiding it.

6. Consistently use `devtools::document()` and make sure `devtools::check()` results in no error.

MaximiliaEggle commented:
Hi Anand, 

thank you for your feedback, I changed the points according to your suggestion, resolved the conflicts, ran it on Terminate/NBE cohort and did a devtools check (no errors or warnings). 

@VerenaK90 I talked to Anand about the .I because I also thought we needed it, but the results are the same and Anand said it is fine if we use the vectorized version.

VerenaK90 reviewed (approved):
Thank you, Maxi! I only made a very small change to the description and will now merge this branch.

VerenaK90 commented:
Hi Maxi, I just came across this tool, which prospectively could be an alternative for signature assignment with the advantage that we could directly use this as an R interface. maybe sth to discuss.

https://github.com/kgori/sigfit
